experiment/adscert/inprocesssigner.go

```diff
@@ -0,0 +1,64 @@
+package adscert
+
+import (
+	"crypto/rand"
+	"errors"
+	"github.com/IABTechLab/adscert/pkg/adscert/api"
+	"github.com/IABTechLab/adscert/pkg/adscert/discovery"
+	"github.com/IABTechLab/adscert/pkg/adscert/signatory"
+	"github.com/benbjohnson/clock"
+	"github.com/prebid/prebid-server/config"
+	"net/url"
+	"time"
+)
+
+// inProcessSigner - holds the signatory to add adsCert header to requests using in process go library
+type inProcessSigner struct {
+	signatory signatory.AuthenticatedConnectionsSignatory
+}
+
+// Sign - adds adsCert header to requests using in process go library
+func (ips *inProcessSigner) Sign(destinationURL string, body []byte) (string, error) {
+	req := &api.AuthenticatedConnectionSignatureRequest{
+		RequestInfo: createRequestInfo(destinationURL, body),
+	}
+	signatureResponse, err := ips.signatory.SignAuthenticatedConnection(req)
+	if err != nil {
+		return "", err
+	}
+	return getSignatureMessage(signatureResponse)
+}
+
+func newInProcessSigner(inProcessSignerConfig config.AdsCertInProcess) (*inProcessSigner, error) {
+	if err := validateInProcessSignerConfig(inProcessSignerConfig); err != nil {
+		return nil, err
+	}
+	return &inProcessSigner{
+		signatory: signatory.NewLocalAuthenticatedConnectionsSignatory(
+			inProcessSignerConfig.Origin,
+			rand.Reader,
+			clock.New(),
+			discovery.NewDefaultDnsResolver(),
+			discovery.NewDefaultDomainStore(),
+			time.Duration(inProcessSignerConfig.DNSCheckIntervalInSeconds)*time.Second,
+			time.Duration(inProcessSignerConfig.DNSRenewalIntervalInSeconds)*time.Second,
+			[]string{inProcessSignerConfig.PrivateKey}),
+	}, nil
+}
+
+func validateInProcessSignerConfig(inProcessSignerConfig config.AdsCertInProcess) error {
+	_, err := url.ParseRequestURI(inProcessSignerConfig.Origin)
```

Comment:
I like the added validation but my initial suggestion for adding validation was actually on the `AdsCertInProcess` config struct. Similar to what you'd see for other config options like for example, `GDPR`, currency conversion, etc over [here](https://github.com/prebid/prebid-server/blob/master/config/config.go#L107). I like it better there since the validation is related to the configuration and so makes sense to have that validation function on the `AdsCertInProcess` config option rather than here. At this point i.e. in `newInProcessSigner`,  we should have valid config and should just be creating a new signer.

Reply:
`IMHO better to trap the errors earliest when the app starts` - this is how it is now: router.go -> func New calls NewAdCertsSigner -> it calls `newInProcessSigner` or `newRemoteSigner` and validation is the first thing we run there. 
I can move validation to config, the only concern is if we want to "spread"  and partly duplicate adscert logic to config. For example: do not validate anything if `!experimentAdCertsConfig.Enabled` and check i which one to validate: `len(experimentAdCertsConfig.InProcess.Origin) > 0 && len(experimentAdCertsConfig.Remote.Url) > 0 `

Reply:
I believe all config validation logic is currently in config.fo, so it moving it there would keep it central

I understand about knowing which side to configure. I didn't focus on that before. Perhaps we need to change Enabled to Mode (or some such, whatever language we use elsewhere) which can be something like "off", "inprocess" or "remote".  Then we know exactly the host's intent on which one to validate, and also perhaps that the other choice has not been configured.